pkgs/development/libraries/vulkan-loader/default.nix

```diff
@@ -14,7 +14,9 @@ stdenv.mkDerivation rec {
     });
 
   nativeBuildInputs = [ cmake pkg-config ];
-  buildInputs = [ libX11 libxcb libXrandr vulkan-headers wayland ];
+  buildInputs = [ vulkan-headers ]
+    ++ lib.optionals (!stdenv.isDarwin) [ libX11 libxcb libXrandr wayland ];
+  propagatedBuildInputs = lib.optional stdenv.isDarwin darwin.moltenvk;
```

Comment:
MoltenVK is an ICD; it should not be a direct dependency of vulkan-loader, for the same reason that e.g. mesa isn't. If a package depends on vulkan-loader, then it relies on vulkan-loader to search the environment at *runtime* to find a suitable ICD. On NixOS this is mediated via `addOpenGLRunpath.driverLink` below. If that doesn't already work on darwin, a similar mechanism should be developed; perhaps instead supply `-DSYSCONFDIR=${moltenvk}/share`, though in the long run a mechanism which admits other ICDs (e.g. llvmpipe) would be preferred.

Reply:
Since MoltenVK is really the only practical option on Darwin, I wasn’t sure. Of the things I’ve tested (Wine, vkQuake), they all link directly against `libMoltenVK.dylib` on Darwin instead of going through vulkan-loader (even when it’s present in the derivation’s `buildInputs`). I tried to find something that built on Darwin that would use vulkan-loader mechanism but didn’t have much luck. The other app I tried (airshipper) ended up not even using Vulkan once I fixed its `buildInputs` for Darwin.

Anyway, thanks for the feedback. I’ll push a change to make MoltenVK honor the intent of vulkan-loader and not propagate the dependency just for convenience.

Reply:
I ended up punting on multiple ICDs. I assume doing so would require coordination with nix-darwin to establish a place for ICD-type stuff (presumably following NixOS’s lead), but Darwin Nix users won’t necessarily be using nix-darwin. It just seems like a lot of additional complexity for a very niche use case (and llvmpipe doesn’t even support aarch64 currently).

Reply:
> Of the things I’ve tested (Wine, vkQuake), they all link directly against libMoltenVK.dylib on Darwin instead of going through vulkan-loader

Should those be depending on vulkan-loader at all, then? Sounds like a direct moltenvk dependency is called for in those cases. vulkan-loader is useful on macos--it allows the validation layers to be used, for example--but if software doesn't use it, it doesn't seem terribly sensible to pretend otherwise.